pkg/utils/runnable_server_test.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"net"
+	"net/http"
+	"testing"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+func NewRunnableServer() (*RunnableServer, error) {
+	l, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return nil, err
+	}
+
+	s := &http.Server{
+		Addr: l.Addr().String(),
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+			w.WriteHeader(http.StatusOK)
+		}),
+	}
+
+	rs := &RunnableServer{
+		Server:    s,
+		ServeFunc: func() error { return s.Serve(l) },
+		Logger:    zap.NewNop(),
+	}
+
+	return rs, nil
+}
+
+func TestRunnableServerWithShutdownTimeout(t *testing.T) {
```

Comment:
Just so I am clear, this isn't testing that there is a timeout on shutdown (i.e. even a misbehaving Shutdown still returns after the timeout). Rather it is testing that closing `stopCh` shutdowns the server. Right?

Reply:
Correct. RunnableServer relies on Shutdown to return when its given context times out, but as you've identified, this test doesn't verify that the context was passed correctly to Shutdown. I'll add a test for that.

Reply:
Test added. I also changed Start to return the Shutdown error, and it occasionally flakes. I'll have to debug more next week.

Reply:
The test is more stable now, so this is RFAL.